### PR TITLE
#29 add cancelreplacemarket

### PR DESCRIPTION
**Pull request by blewater**
Rename CancelReplaceOrder-> CancelReplaceLimitOrder when it was still left as CancelReplaceOrder i.e. keeper, unit test.
Add CancelReplaceMarketOrder + tests.

Please note, you'll see changes from the previous open PR #30 here.

### CODE REVIEW

File: x/market/keeper/keeper.go [hunk 1]
```diff
@@ -417,6 +419,61 @@ func (k *Keeper) CancelReplaceOrder(ctx sdk.Context, newOrder types.Order, origC
 	return &sdk.Result{Events: evts}, nil
 }
 
+func (k *Keeper) CancelReplaceMarketOrder(ctx sdk.Context, msg *types.MsgCancelReplaceMarketOrder) (*sdk.Result, error) {
+	// Use a fixed gas amount
+	ctx.GasMeter().ConsumeGas(gasPriceCancelReplaceOrder, "CancelReplaceMarketOrder")
```
Comment (haasted) [inline x/market/keeper/keeper.go:424]: This function has quite a lot in common with `CancelReplaceLimitOrder`, ie. checking preconditions, finding original order, verifying that instruments remain the same. Would it be possible to capture that common logic in an internal, shared function?

File: x/market/keeper/keeper_test.go [hunk 2]
```diff
@@ -226,6 +226,179 @@ func TestMarketOrderSlippage1(t *testing.T) {
...
+	foundOrder := k.GetOrderByOwnerAndClientOrderId(
+		ctx, acc1.GetAddress().String(), clientID,
+	)
+	require.NotNil(t, foundOrder, "Market order should exist")
+	// 100% slippage to double source
+	require.True(t, foundOrder.Source.IsEqual(sdk.NewCoin("gbp", sdk.NewInt(200))))
+
+	// Gave 1 gbp and gained a eur
+	acc1Bal := bk.GetAllBalances(ctx, acc1.GetAddress())
+	require.Equal(t, coins("1eur,499gbp").String(), acc1Bal.String())
```
Comment (haasted) [inline x/market/keeper/keeper_test.go:269]: Isn't it possible to do an `Equal` on two `coins` directly?

File: x/market/keeper/keeper.go [hunk 3]
```diff
@@ -417,6 +419,61 @@ func (k *Keeper) CancelReplaceOrder(ctx sdk.Context, newOrder types.Order, origC
 	return &sdk.Result{Events: evts}, nil
 }
 
+func (k *Keeper) CancelReplaceMarketOrder(ctx sdk.Context, msg *types.MsgCancelReplaceMarketOrder) (*sdk.Result, error) {
```
Comment (haasted) [inline x/market/keeper/keeper.go:422]: The pattern is usually to not send `Msg` objects into the Keeper, but I can see why it's necessary in this case. There are so many parameters that need to be handled.

Could you evaluate the following approach, and see if it results in a better structure and avoids the `Msg` object in the Keeper:
1. Change `NewMarketOrderWithSlippage` to NOT submit the order to the book, but merely calculate the price and return a new `Order` instance.
2. Use that function in the message handler to create an Order object that can then be submitted to `CancelReplaceMarketOrder`.

It's possible that I overlooked something, and the above is untenable. :)

File: x/market/keeper/keeper_test.go (hunk 2 above)
Comment (blewater) [reply x/market/keeper/keeper_test.go:269]: While it is, by stringifying it displays the comparison result clearly in decimal vs hex.

File: x/market/keeper/keeper.go (hunk 3 above)
Comment (blewater) [reply x/market/keeper/keeper.go:422]: To keep the Keeper API symmetric the same principle can be applied to the AddLimit / MarketOrder and will reuse the same price formula. Shall I?

Comment (haasted) [reply x/market/keeper/keeper.go:422]: Yes, I think it should probably apply to all cases.

**Comment by blewater** [comment]
This PR also includes a refactoring of the slippage market orders to be issued as limit orders at the Message server layer vs the keeper layer.

File: x/market/keeper/keeper.go [hunk 4]
```diff
@@ -114,23 +110,49 @@ func (k *Keeper) createExecutionPlan(ctx sdk.Context, SourceDenom, DestinationDe
...
+// (dst amount/last market price) and adding the slippage percentage is the
+// resulting value. The maxSlippage decimal expresses a percentage (1 is 100%).
+func (k *Keeper) GetSrcFromSlippage(
+	ctx sdk.Context, srcDenom string, dst sdk.Coin, maxSlippage sdk.Dec,
+) (sdk.Coin, error) {
+	if err := sdk.ValidateDenom(srcDenom); err != nil {
+		return sdk.Coin{}, sdkerrors.Wrapf(types.ErrInvalidInstrument,"%s", err)
+	}
+
+	if !dst.IsValid() {
```
Comment (haasted) [inline x/market/keeper/keeper.go:124]: This check is already performed in the `ValidateBasic` method of the `Msg`. It's ok to assume that basic validation like that has taken place before the keeper logic.

File: x/market/keeper/keeper.go [hunk 5]
```diff
@@ -276,11 +298,11 @@ func (k *Keeper) NewOrderSingle(ctx sdk.Context, aggressiveOrder types.Order) (*
 			nextSourceFilledCoin := sdk.NewCoin(passiveOrder.Source.Denom, stepSourceFilled.RoundInt())
 			err := k.transferTradedAmounts(ctx, nextDestinationFilledCoin, nextSourceFilledCoin, passiveOrder.Owner, aggressiveOrder.Owner)
 			if err != nil {
-				panic(err)
+				fmt.Println(nextDestinationFilledCoin, nextSourceFilledCoin)
```
Comment (haasted) [inline x/market/keeper/keeper.go:301]: Use logging instead of `Println`. Error should probably be returned.

**Review by haasted — APPROVED** [review]
Only minor comments. Feel free to merge.

File: x/market/keeper/keeper.go (hunk 5 above)
Comment (blewater) [reply x/market/keeper/keeper.go]: :+1: yeah debugging slipped, i reverted to the original

**Pull request merged**